kafka/integration_test.go

```diff
@@ -0,0 +1,529 @@
+
+	default:
+		fmt.Printf("Ignored event: %s\n", e)
+	}
+}
+
+// TestConsumerFuncBasedAPI test function-based API, requires broker
+func TestConsumerFuncBasedAPI(t *testing.T) {
+	runConsumerFuncBasedAPI(t, "CommitMessage")
+	runConsumerFuncBasedAPI(t, "CommitOffsets")
+	runProducerChannelBasedAPI(t)
+	runConsumerFuncBasedAPI(t, "Commit")
+}
+
+func runConsumerFuncBasedAPI(t *testing.T, commitMode string) {
+	if !testconfRead() {
+		t.Skipf("Missing testconf.json")
+	}
+	c, err := NewConsumer(&ConfigMap{
+		"bootstrap.servers":    testconf.Brokers,
+		"group.id":             groupId,
+		"enable.auto.commit":   false,
+		"session.timeout.ms":   6000,
+		"default.topic.config": ConfigMap{"auto.offset.reset": "earliest"}})
+
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "Failed to create consumer: %s\n", err)
+		os.Exit(1)
+	}
+
+	t.Logf("Created Consumer %v\n", c)
+
+	err = c.Subscribe(testconf.Topic, nil)
+
+	maxMsgCnt := 3
+	msgcnt := 0
+	run := true
+
+	savedMessages := make([]*Message, maxMsgCnt)
+
+	for run == true {
+		select {
+		default:
+			ev := c.Poll(100)
+			if ev == nil {
+				continue
+			}
+
+			switch e := ev.(type) {
+			case *Message:
+				fmt.Printf("%% Message on %s: key=(%s), value=(%s)\n",
+					e.TopicPartition, e.Key, e.Value)
+				savedMessages[msgcnt] = e
+				msgcnt++
+				if msgcnt >= maxMsgCnt {
+					run = false
+				}
+			case PartitionEOF:
+				fmt.Printf("%% Reached %v\n", e)
+				run = false
+			case Error:
+				fmt.Fprintf(os.Stderr, "%% Error: %v\n", e)
+				run = false
+			default:
+				fmt.Printf("Ignored %v\n", e)
+			}
+		}
+	}
+
+	switch commitMode {
+	case "CommitMessage":
+		// verify CommitMessage() API
+		for _, message := range savedMessages {
+			_, commit_err := c.CommitMessage(message)
+			if commit_err != nil {
+				t.Errorf("Cannot commit message\n")
+			}
+		}
+	case "CommitOffsets":
+		// verify CommitOffset
+		partitions := make([]TopicPartition, len(savedMessages))
+		for index, message := range savedMessages {
+			partitions[index] = message.TopicPartition
+		}
+		_, commit_err := c.CommitOffsets(partitions)
+		if commit_err != nil {
+			t.Errorf("Failed to commit using CommitOffsets\n")
+		}
+	default:
+		// verify Commit() API
+		_, commit_err := c.Commit()
+		if commit_err != nil {
+			t.Errorf("Failed to commit: %s", commit_err)
+		}
+
+	}
+
+	t.Logf("Closing consumer\n")
+	c.Close()
+}
+
+// TestConsumerChannelBasedAPI test channel-based API, requires broker
+func TestConsumerChannelBasedAPI(t *testing.T) {
+	if !testconfRead() {
+		t.Skipf("Missing testconf.json")
+	}
+	c, err := NewConsumer(&ConfigMap{
+		"bootstrap.servers":        testconf.Brokers,
+		"group.id":                 groupId,
+		"enable.auto.commit":       false,
+		"session.timeout.ms":       6000,
+		"go.events.channel.enable": true,
+		"default.topic.config":     ConfigMap{"auto.offset.reset": "earliest"}})
+
+	if err != nil {
+		t.Fatalf("Failed to create consumer: %s\n", err)
+	}
+
+	t.Logf("Created Consumer %v\n", c)
```

Comment:
Could we avoid code duplication (and provide some verification in this one) by reusing the code above from runCon..?